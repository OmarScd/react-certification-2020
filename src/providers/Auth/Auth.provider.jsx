import React, { useState, useEffect, useContext } from 'react';

import { USERNAME_STORAGE_KEY } from '../../utils/constants';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error(`Can't use "useAuth" without an AuthProvider!`);
  }
  return context;
};

const AuthProvider = ({ children }) => {
  const [authenticated, setAuthenticated] = useState(false);
  const [username, setUsername] = useState(undefined);

  useEffect(() => {
    const user = storage.get(USERNAME_STORAGE_KEY);
    const isAuthenticated = !!user;

    setAuthenticated(isAuthenticated);
    if (isAuthenticated) {
      setUsername(user);
    }
  }, []);

  const login = (name) => {
    setAuthenticated(true);
    setUsername(name);
    storage.set(USERNAME_STORAGE_KEY, name);
  };

  const logout = () => {
    setAuthenticated(false);
    setUsername(undefined);
    storage.remove(USERNAME_STORAGE_KEY);
  };

  return (
    <AuthContext.Provider value={{ login, logout, authenticated, username }}>
      {children}
    </AuthContext.Provider>
  );
};

export { useAuth, AuthProvider as default };
