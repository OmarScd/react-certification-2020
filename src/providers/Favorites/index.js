export { default } from './Favorites.provider';
