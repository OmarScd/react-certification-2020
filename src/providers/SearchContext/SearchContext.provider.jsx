import React, { useState, useContext } from 'react';

const SearchContext = React.createContext();

const useSearch = () => {
  const context = useContext(SearchContext);
  if (!context) {
    throw new Error(`Can't use "useSearch" without a FavoritesProvider!`);
  }
  return context;
};

const SearchContextProvider = ({ children }) => {
  const [query, setQuery] = useState('Subdivisions');

  return (
    <SearchContext.Provider value={{ query, setQuery }}>
      {children}
    </SearchContext.Provider>
  );
};

export { useSearch, SearchContextProvider as default };
