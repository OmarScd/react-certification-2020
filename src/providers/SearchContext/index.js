export { default, useSearch } from './SearchContext.provider';
