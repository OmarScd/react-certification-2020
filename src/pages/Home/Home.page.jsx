import React, { useContext, useEffect, useState } from 'react';
import styled from 'styled-components';

import Layout from '../../components/Layout';
import SearchContext from '../../contexts/SearchContext';
import VideoCard from '../../components/VideoCard';
import { getQueryURL, fetchData } from '../../utils/fns';
import useDebounce from '../../utils/hooks/useDebounce';

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin: 0 auto;
  max-width: 1500px;
  padding: 0 3rem;
  justify-content: center;
`;

function HomePage() {
  const { query } = useContext(SearchContext);

  const debouncedQuery = useDebounce(query, 400);

  const [searchResult, setResult] = useState(null);

  useEffect(() => {
    fetchData(getQueryURL(debouncedQuery)).then((data) => setResult(data));
  }, [debouncedQuery]);

  return (
    <Layout>
      {searchResult ? (
        <Container>
          {searchResult.items.map((item) => (
            <VideoCard
              key={item.id.videoId}
              id={item.id.videoId}
              thumbnail={item.snippet.thumbnails.medium.url}
              title={item.snippet.title}
              description={item.snippet.description}
            />
          ))}
        </Container>
      ) : (
        <div style={{ textAlign: 'center', padding: '2rem' }}>Loading...</div>
      )}
    </Layout>
  );
}

export default HomePage;
