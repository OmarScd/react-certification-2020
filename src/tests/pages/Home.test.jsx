import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import Home from '../../pages/Home';
import ThemeProvider from '../../providers/Theme';
import '../../providers/SearchContext';
import '../../providers/Auth';

jest.mock('../../providers/SearchContext', () => ({
  default: jest.fn(),
  useSearch: () => ({ query: 'test', setQuery: jest.fn() }),
}));

jest.mock('../../providers/Auth', () => ({
  default: jest.fn(),
  useAuth: () => ({ authenticated: true }),
}));

jest.mock('../../utils/fns.js', () => {
  const fetchData = jest.fn(() => {
    // eslint-disable-next-line global-require
    const apiSearchResponse = require('../fixtures/apiSearchResponse');
    return Promise.resolve(apiSearchResponse[0]);
  });
  fetchData.mockReturnValueOnce(Promise.resolve(false));

  return {
    getQueryURL: jest.fn(),
    getRelatedURL: jest.fn(),
    getInfoURL: jest.fn(),
    fetchData,
  };
});

describe('Home page', () => {
  afterAll(cleanup);

  it('renders correctly without api response', () => {
    const { container } = render(
      <ThemeProvider>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </ThemeProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with api response', () => {
    const { container } = render(
      <ThemeProvider>
        <BrowserRouter>
          <Home />
        </BrowserRouter>
      </ThemeProvider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
