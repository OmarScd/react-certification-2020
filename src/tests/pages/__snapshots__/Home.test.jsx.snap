// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page renders correctly with api response 1`] = `
<div
  class="container"
>
  <header
    class="sc-hKFyIo jcdTAl"
  >
    <div
      class="sc-eCApGN hCdAiT"
    >
      <div
        class="sc-jSFkmK qmXvl"
      >
        <a
          class="sc-gKAblj iUhdyg"
          href="/"
        >
          Home
        </a>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <a
          class="sc-gKAblj iUhdyg"
          href="/favorites"
        >
          Favorites
        </a>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <form
          class="sc-bdnylx hYwFxl"
        >
          <img
            alt="magnifying lens"
            class="sc-dlnjPT dGgzTM"
            src="search-24px.svg"
          />
          <input
            class="sc-gtssRu fkigkU"
            placeholder="Search..."
            type="text"
            value="test"
          />
        </form>
      </div>
    </div>
    <div
      class="sc-eCApGN hCdAiT"
    >
      <div
        class="sc-jSFkmK qmXvl"
      >
        <button
          class="sc-iCoHVE kHdmrf"
        >
          Log out
        </button>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <img
          alt="person"
          class="sc-fujyUd hfIOzZ"
          src="person-24px.svg"
        />
      </div>
    </div>
  </header>
  <div
    class="content"
  >
    <div
      style="text-align: center; padding: 2rem;"
    >
      Loading...
    </div>
  </div>
</div>
`;

exports[`Home page renders correctly without api response 1`] = `
<div
  class="container"
>
  <header
    class="sc-hKFyIo jcdTAl"
  >
    <div
      class="sc-eCApGN hCdAiT"
    >
      <div
        class="sc-jSFkmK qmXvl"
      >
        <a
          class="sc-gKAblj iUhdyg"
          href="/"
        >
          Home
        </a>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <a
          class="sc-gKAblj iUhdyg"
          href="/favorites"
        >
          Favorites
        </a>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <form
          class="sc-bdnylx hYwFxl"
        >
          <img
            alt="magnifying lens"
            class="sc-dlnjPT dGgzTM"
            src="search-24px.svg"
          />
          <input
            class="sc-gtssRu fkigkU"
            placeholder="Search..."
            type="text"
            value="test"
          />
        </form>
      </div>
    </div>
    <div
      class="sc-eCApGN hCdAiT"
    >
      <div
        class="sc-jSFkmK qmXvl"
      >
        <button
          class="sc-iCoHVE kHdmrf"
        >
          Log out
        </button>
      </div>
      <div
        class="sc-jSFkmK qmXvl"
      >
        <img
          alt="person"
          class="sc-fujyUd hfIOzZ"
          src="person-24px.svg"
        />
      </div>
    </div>
  </header>
  <div
    class="content"
  >
    <div
      style="text-align: center; padding: 2rem;"
    >
      Loading...
    </div>
  </div>
</div>
`;
